genkit-tools/common/src/eval/localFileDatasetStore.ts

```diff
@@ -0,0 +1,221 @@
+    if (!req.datasetId) {
+      return this.createDatasetInternal(req.patch, req.displayName);
+    }
+    const datasetId = req.datasetId;
+    const filePath = path.resolve(
+      this.storeRoot,
+      this.generateFileName(datasetId)
+    );
+    if (!fs.existsSync(filePath)) {
+      return this.createDatasetInternal(req.patch, req.displayName);
+    }
+
+    const metadatas = await this.listDatasets();
+    const prevMetadata = metadatas.find(
+      (metadata) => metadata.datasetId === datasetId
+    );
+    if (!prevMetadata) {
+      throw new Error(`Update dataset failed: dataset metadata not found`);
+    }
+
+    logger.info(`Updating Dataset at ` + filePath);
+    await writeFile(filePath, JSON.stringify(req.patch));
+
+    const newMetadata = {
+      datasetId: datasetId,
+      size: Array.isArray(req.patch)
+        ? req.patch.length
+        : req.patch.samples.length,
+      uri: 'file://' + filePath,
+      version: prevMetadata.version++,
+      displayName: req.displayName,
+      createTime: prevMetadata.createTime,
+      updateTime: Date.now().toString(),
+    };
+
+    logger.debug(
+      `Updating DatasetMetadata for ID ${datasetId} at ` +
+        path.resolve(this.indexFile)
+    );
+    const prevMetadataIndex = metadatas.indexOf(prevMetadata);
+    // Replace the metadata object with the new metadata
+    metadatas[prevMetadataIndex] = newMetadata;
+    await writeFile(
+      path.resolve(this.indexFile),
+      metadatas
+        .map((metadata) => JSON.stringify(metadata))
+        .join(this.INDEX_DELIMITER)
+    );
+
+    return newMetadata;
+  }
+
+  async getDataset(id: string): Promise<Dataset> {
+    const filePath = path.resolve(this.storeRoot, this.generateFileName(id));
+    if (!fs.existsSync(filePath)) {
+      throw new Error(`Dataset not found for dataset ID {$id}`);
+    }
+    return await readFile(filePath, 'utf8').then((data) =>
+      EvalFlowInputSchema.parse(JSON.parse(data))
+    );
+  }
+
+  async listDatasets(): Promise<DatasetMetadata[]> {
+    let metadatas = await readFile(this.indexFile, 'utf8').then((data) => {
+      if (!data) {
+        return [];
+      }
+      // strip the final carriage return before parsing all lines
+      return data
+        .slice(0, -1)
+        .split(this.INDEX_DELIMITER)
+        .map(this.parseLineToDatasetMetadata);
+    });
+
+    logger.debug(`Found Dataset Metadatas: ${JSON.stringify(metadatas)}`);
+
+    return metadatas;
+  }
+
+  async deleteDataset(id: string): Promise<void> {
+    const filePath = path.resolve(this.storeRoot, this.generateFileName(id));
+    if (!fs.existsSync(filePath)) {
+      return;
+    }
+    return await rm(filePath);
+  }
+
+  private generateRootPath(): string {
+    const rootHash = crypto
+      .createHash('md5')
+      .update(process.cwd() || 'unknown')
+      .digest('hex');
+    return path.resolve(os.tmpdir(), `.genkit/${rootHash}/datasets`);
```

Comment:
We should not store under `tmpdir`. Use `process.cwd()` instead?

Reply:
Done. This is harder to mock. So, I have used regular expression matching in tests and also got rid of the URI, since it not required currently anyway.